### PR TITLE
feature/date.js-extension

### PR DESCRIPTION
**Pull request by yehor-podporinov**

### CODE REVIEW

File: packages/utils/src/date.ts [hunk 1]
```diff
@@ -8,12 +8,24 @@ import dayjs, {
 } from 'dayjs'
 import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
 import isSameOrBefore from 'dayjs/plugin/isSameOrBefore'
-import { IsoDate, UnixDate } from '@/types'
+import calendar from 'dayjs/plugin/calendar'
```
Comment (ArtSuslov) [inline packages/utils/src/date.ts]: import { calendar, isSameOrAfter, isSameOrBefore  } from 'dayjs/plugin'

File: packages/utils/src/date.ts [hunk 2]
```diff
@@ -8,12 +8,24 @@ import dayjs, {
...
+import { IsoDate, UnixDate, StringDate } from '@/types'
+
+const INPUT_DATE_FORMAT: OptionType = 'DD/MM/YYYY'
 
 export class DateUtil {
+  static get ISOFormat(): IsoDate {
+    return 'YYYY-MM-DDT00:00:00+00:00'
```
Comment (ArtSuslov) [inline packages/utils/src/date.ts]: are u sure there shouldn't be a trailing 'Z'  to show the time-zone?

**Review by ArtSuslov — CHANGES_REQUESTED** [review]

File: packages/utils/src/date.ts (hunk 1 above)
Comment (yehor-podporinov) [reply packages/utils/src/date.ts]: It's not working

File: packages/utils/src/date.ts (hunk 2 above)
Comment (yehor-podporinov) [reply packages/utils/src/date.ts]: I've removed this getter, because it not used in web client in tokend

File: packages/utils/README.md [hunk 3]
```diff
@@ -12,6 +12,27 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+<details><summary>1.2.0</summary>
```
Comment (ArtSuslov) [inline packages/utils/README.md]: 1.2.0 -> Unreleased

File: packages/utils/package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@distributedlab/utils",
   "description": "Typescript-base utils for Distributed Lab projects",
-  "version": "1.1.0",
+  "version": "1.2.0",
```
Comment (ArtSuslov) [inline packages/utils/package.json]: 1.1.0

**Review by ArtSuslov — CHANGES_REQUESTED** [review]

File: packages/utils/README.md (hunk 3 above)
Comment (yehor-podporinov) [reply packages/utils/README.md]: Done

File: packages/utils/package.json (hunk 4 above)
Comment (yehor-podporinov) [reply packages/utils/package.json]: Done

File: packages/utils/src/date.ts [hunk 5]
```diff
@@ -8,28 +8,132 @@ import dayjs, {
...
+        case 6:
+          return `Прошлая ${this.format(format).toLowerCase()}`
+        default:
+          return `Прошлый ${this.format(format).toLowerCase()}`
+      }
+    },
+    nextWeek: 'dddd H:mm',
+    sameElse: 'D MMMM YYYY HH:mm',
+  },
+})
```
Comment (napalmpapalam) [inline packages/utils/src/date.ts]: create method in DateUtils to update locales outside of the package

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: packages/utils/src/time.ts [hunk 6]
```diff
@@ -2,19 +2,37 @@ import dayjs, { Dayjs } from 'dayjs'
 import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
 import isSameOrBefore from 'dayjs/plugin/isSameOrBefore'
 import relativeTime from 'dayjs/plugin/relativeTime'
+import isBetween from 'dayjs/plugin/isBetween'
+import calendar from 'dayjs/plugin/calendar'
+import utc from 'dayjs/plugin/utc'
+import customParseFormat from 'dayjs/plugin/customParseFormat'
+import updateLocale from 'dayjs/plugin/updateLocale'
+import timezone from 'dayjs/plugin/timezone'
+import duration, { Duration } from 'dayjs/plugin/duration'
```
Comment (napalmpapalam) [inline packages/utils/src/time.ts]: create own one Duration type

File: packages/utils/src/time.ts [hunk 7]
```diff
@@ -43,10 +75,18 @@ export class Time {
...
+  public get ms(): UnixDate {
+    return this.#date.valueOf()
+  }
+
   public get ISO(): IsoDate {
     return this.#date.toISOString()
   }
 
+  public get RFC3339(): IsoDate {
+    return this.#date.utc(true).format('YYYY-MM-DDTHH:mm:ss') + 'Z'
```
Comment (napalmpapalam) [inline packages/utils/src/time.ts]:
```ts
format('YYYY-MM-DDTHH:mm:ss[Z]')
```

File: packages/utils/src/time.ts [hunk 8]
```diff
@@ -123,7 +185,45 @@ export class Time {
...
+    isLocal?: boolean,
+  ): string {
+    return dayjs.locale(preset, object, isLocal)
+  }
+
+  public static setDefaultTimezone(timezone?: string): void {
+    dayjs.tz.setDefault(timezone)
+  }
+
+  public static updateLocale(
```
Comment (napalmpapalam) [inline packages/utils/src/time.ts]: `setLocale`

File: packages/utils/src/time.ts [hunk 9]
```diff
@@ -123,7 +185,45 @@ export class Time {
   public get toNow(): string {
     return this.#date.toNow()
   }
+
+  public static locale(
```
Comment (napalmpapalam) [inline packages/utils/src/time.ts:188]: do u tested it?

File: packages/utils/src/time.ts [hunk 10]
```diff
@@ -123,7 +185,45 @@ export class Time {
...
+
+  public static updateLocale(
+    localeName: string,
+    customConfig: TimeLocale,
+  ): TimeLocale {
+    return dayjs.updateLocale(localeName, customConfig)
+  }
+}
+
+export class TimeDuration {
```
Comment (napalmpapalam) [inline packages/utils/src/time.ts]: rename to `Duration`, why it's not in the separated file?

File: packages/utils/src/time.ts [hunk 11]
```diff
@@ -123,7 +185,45 @@ export class Time {
...
+    this.#duration = this._duration(units)
+  }
+
+  private _duration(units: TimeDurationUnitsObject): Duration {
+    return dayjs.duration(units)
+  }
+
+  public asMilliseconds(): UnixDate {
+    return this.#duration.asMilliseconds()
+  }
```
Comment (napalmpapalam) [inline packages/utils/src/time.ts]: add more common methods

File: packages/utils/src/types/time.ts [hunk 12]
```diff
@@ -42,3 +44,68 @@ export type TimeUnit = TimeUnitLong | TimeUnitLongPlural | TimeUnitShort
...
+// export type DurationUnits = Partial<{
+//   milliseconds: number
+//   seconds: number
+//   minutes: number
+//   hours: number
+//   days: number
+//   months: number
+//   years: number
+//   weeks: number
+// }>
```
Comment (napalmpapalam) [inline packages/utils/src/types/time.ts]: ?????????/

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: packages/utils/src/date.ts (hunk 2 above)
Comment (napalmpapalam) [reply packages/utils/src/date.ts]: it doesn't matter, not only tokend uses this package

File: packages/utils/README.md [hunk 13]
```diff
@@ -12,6 +12,27 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+<details><summary>Unreleased</summary>
+  <h4>Added</h4>
+    <h5>New getters to Time</h5>
```
Comment (napalmpapalam) [inline packages/utils/README.md]: if u add something it's obvious that it's new

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: packages/utils/src/duration.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  public asWeeks(): UnixDate {
+    return this.#duration.asWeeks()
+  }
+  public asMonths(): UnixDate {
+    return this.#duration.asMonths()
+  }
+  public asYears(): UnixDate {
+    return this.#duration.asYears()
+  }
+  public humanize(): string {
```
Comment (napalmpapalam) [inline packages/utils/src/duration.ts]: make all methods as getters without `as` prefix

File: packages/utils/src/duration.ts [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    return this.#duration.humanize()
+  }
+}
+
+export const timeDuration = (units: TimeDurationUnitsObject): TimeDuration =>
```
Comment (napalmpapalam) [inline packages/utils/src/duration.ts]: `duration`

File: packages/utils/src/duration.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+  constructor(units: TimeDurationUnitsObject) {
+    this.#duration = this._duration(units)
+  }
+
+  private _duration(units: TimeDurationUnitsObject): Duration {
+    return dayjs.duration(units)
+  }
+
+  public asMilliseconds(): UnixDate {
```
Comment (napalmpapalam) [inline packages/utils/src/duration.ts]: return type is number, not `UnixDate`, unix date is timestamp

File: packages/utils/src/time.ts (hunk 10 above)
Comment (napalmpapalam) [reply packages/utils/src/time.ts]: class not renamed

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

**Review by napalmpapalam — APPROVED** [review]

**Pull request merged**